src/sage/schemes/elliptic_curves/ell_point.py

```diff
@@ -1944,21 +1974,17 @@ def tate_pairing(self, Q, n, k, q=None):
             else:
                 raise ValueError("Unexpected field degree: set keyword argument q equal to the size of the base field (big field is GF(q^%s))." % k)
 
-        if n*P != E(0):
-            raise ValueError('This point is not of order n=%s' % n)
-
-        # In small cases, or in the case of pairing an element with
-        # itself, Q could be on one of the lines in the Miller
-        # algorithm. If this happens we try again, with an offset of a
-        # random point.
-        try:
-            ret = self._miller_(Q, n)
-            e = Integer((q**k - 1)/n)
-            ret = ret**e
-        except (ZeroDivisionError, ValueError):
-            R = E.random_point()
-            ret = self.tate_pairing(Q + R, n, k)/self.tate_pairing(R, n, k)
-        return ret
+        if pari.ellmul(E, P, n) != [0]:
+            raise ValueError("The point P must be in the n-torsion")
+
+        # NOTE: Pari returns the non-reduced Tate pairing, so we
+        # must perform the exponentation ourselves using the supplied
+        # k value
+        ePQ = pari.elltatepairing(E, P, Q, n)
+        # TODO: if n or k is chosen badly, this could error, should we
+        # handle this explicitly by ensuring n divides q^k - 1?
+        exp = Integer((q**k - 1)/n)
```

Comment:
I think it is a good idea to check that n divides q^k-1, for example by computing mod(q,n)^k, but only if q has been explicitly supplied and with a suitable error message if the condition fails.  So do this test in the code block which computes q perhaps.